lib/onig-lib.js

```diff
@@ -0,0 +1,184 @@
+          recursionLimit: 5,
+          // Oniguruma option for `^`->`\A`, `$`->`\Z`; improves search performance without any
+          // change in meaning since TM grammars search line by line
+          singleline: true
+        }
+      })
+      cache.set(pattern, regex)
+      return regex
+    } catch (error_) {
+      const error =
+        error_ instanceof Error
+          ? error_
+          : new Error('Unexpected error', {cause: error_})
+      cache.set(pattern, error)
+      if (forgiving) {
+        return null
+      }
+
+      throw error
+    }
+  })
+
+  return {
+    findNextMatchSync(onigString, startPosition) {
+      const string =
+        typeof onigString === 'string' ? onigString : onigString.content
+      /** @type {Array<[index: number, match: RegExpExecArray]>} */
+      const pending = []
+
+      for (const [index, regexp] of regexps.entries()) {
+        if (!regexp) {
+          continue
+        }
+
+        try {
+          regexp.lastIndex = startPosition
+          const match = regexp.exec(string)
+
+          if (!match) {
+            continue
+          }
+
+          // If the match is at the start position, return it immediately
+          if (match.index === startPosition) {
+            return toResult(index, match)
+          }
+
+          // Otherwise, store it for later
+          pending.push([index, match])
+        } catch (error) {
+          if (!forgiving) {
+            throw error
+          }
+        }
+      }
+
+      // Find the closest match to the start position
+      if (pending.length > 0) {
+        const minIndex = Math.min(...pending.map((match) => match[1].index))
+        for (const [index, match] of pending) {
+          if (match.index === minIndex) {
+            return toResult(index, match)
+          }
+        }
+      }
+
+      return null
+    },
+    dispose: noop
+  }
+}
+
+/**
+ * Create an oniguruma string object compatible with `vscode-textmate`.
+ *
+ * @param {string} content
+ *   The string content to wrap.
+ * @returns {OnigString}
+ *   The oniguruma string object.
+ */
+export function createOnigString(content) {
+  return {
+    content,
+    dispose: noop
+  }
+}
+
+/**
+ * Do nothing
+ *
+ * @returns {undefined}
+ */
+function noop() {}
+
+/**
+ * Convert a RegExp match to an oniguruma match.
+ *
+ * @param {number} index
+ *   The index of the match.
+ * @param {RegExpExecArray} match
+ *   The actual match
+ * @returns {IOnigMatch}
+ *   The oniguruma match.
+ */
+function toResult(index, match) {
+  return {
+    index,
+    captureIndices:
+      match.indices?.map((indice) => {
+        if (indice === undefined) {
+          return {
+            start: MAX,
+            end: MAX,
+            length: 0
+          }
+        }
+
+        return {
+          start: indice[0],
+          end: indice[1],
+          length: indice[1] - indice[0]
+        }
+      }) || []
```

Comment:
- can you sort the fields of these 3 objects
- seems weird to use this max value, is it the last index into some source string? Could that value be passed in as a parameter?